# Домашнее задание к лекции "Базовые понятия статистики"

Обязательная часть  
Будем осуществлять работу с непростым набором данных о состоянии здоровья лошадей, испытывающих кишечные колики.  

Задание 1. Базовое изучение  
Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.  

Задание 2. Работа с выбросами  
В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.  

Задание 3. Работа с пропусками  
Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [1]:
import pandas as pd
import numpy as np

In [2]:
horse_data = pd.read_csv('horse_data.csv')

In [3]:
# Что бы присвоить имена столбцам, прибегнул к помощи экселя
columns_file = pd.read_excel('columns.xlsx')['Normal name']
columns_list = []
for column in columns_file:
    columns_list.append(column)
horse_data.columns = columns_list

In [4]:
names = open('horse_data.names', 'r')

In [44]:
for i in names:
    print(i)

1. TItle: Horse Colic database



2. Source Information

   -- Creators: Mary McLeish & Matt Cecile

	  	Department of Computer Science

		University of Guelph

		Guelph, Ontario, Canada N1G 2W1

		mdmcleish@water.waterloo.edu

   -- Donor:    Will Taylor (taylor@pluto.arc.nasa.gov)

   -- Date:     8/6/89



3. Past Usage:

   -- Unknown



4. Relevant Information:



   -- 2 data files

      -- horse-colic.data: 300 training instances

      -- horse-colic.test: 68 test instances

   -- Possible class attributes: 24 (whether lesion is surgical)

     -- others include: 23, 25, 26, and 27

   -- Many Data types: (continuous, discrete, and nominal)



5. Number of Instances: 368 (300 for training, 68 for testing)



6. Number of attributes: 28



7. Attribute Information:



  1:  surgery?

          1 = Yes, it had surgery

          2 = It was treated without surgery



  2:  Age

          1 = Adult horse

          2 = Young (< 6 months)



  3:  Hospital Number

          - numer

In [4]:
horse_data.head()

surgery?  Age  Hospital Number rectal temperature pulse respiratory rate  \
0        1    1           534817               39.2    88               20   
1        2    1           530334              38.30    40               24   
2        1    9          5290409              39.10   164               84   
3        2    1           530255              37.30   104               35   
4        2    1           528355                  ?     ?                ?   

  temperature of extremities peripheral pulse mucous membranes  \
0                          ?                ?                4   
1                          1                1                3   
2                          4                1                6   
3                          ?                ?                6   
4                          2                1                3   

  capillary refill time  ... packed cell volume total protein  \
0                     1  ...                 50            85   
1                     1  ...              33.00          6.70   
2                     2  ...              48.00          7.20   
3                     2  ...              74.00          7.40   
4                     1  ...                  ?             ?   

  abdominocentesis appearance abdomcentesis total protein outcome  \
0                           2                           2       3   
1                           ?                           ?       1   
2                           3                        5.30       2   
3                           ?                           ?       2   
4                           ?                           ?       1   

  surgical lesion? type of lesion 1 type of lesion 2 type of lesion 3 cp_data  
0                2             2208                0                0       2  
1                2                0                0                0       1  
2                1             2208                0                0       1  
3                2             4300                0                0       2  
4                2                0                0                0       2  

[5 rows x 28 columns]

Столбцы для анализа:
1. surgery?
2. Age
3. rectal temperature
4. pulse
5. peripheral pulse
6. temperature of extremities
7. capillary refill time
8. packed cell volume

In [5]:
horse_data_select = horse_data[['surgery?', 
                                'Age', 
                                'rectal temperature', 
                                'pulse', 
                                'peripheral pulse', 
                                'temperature of extremities',
                                'capillary refill time',
                                'packed cell volume']]

In [6]:
horse_data_select.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   surgery?                    299 non-null    object
 1   Age                         299 non-null    int64 
 2   rectal temperature          299 non-null    object
 3   pulse                       299 non-null    object
 4   peripheral pulse            299 non-null    object
 5   temperature of extremities  299 non-null    object
 6   capillary refill time       299 non-null    object
 7   packed cell volume          299 non-null    object
dtypes: int64(1), object(7)
memory usage: 18.8+ KB


In [7]:
horse_data_select = horse_data_select.replace('?', np.nan) # замена "?" на NaN

In [8]:
# изменение типа данных для расчета стандартных метрик
horse_data_select[['rectal temperature', 
                   'pulse',
                   'temperature of extremities', 
                   'capillary refill time',
                  'packed cell volume']] = horse_data_select[['rectal temperature', 
                   'pulse', 
                   'temperature of extremities', 
                   'capillary refill time',
                  'packed cell volume']].astype('float')

In [9]:
horse_data_select.describe()

Age  rectal temperature       pulse  temperature of extremities  \
count  299.00000          239.000000  275.000000                  243.000000   
mean     1.64214           38.166527   71.934545                    2.345679   
std      2.17730            0.733508   28.680522                    1.046369   
min      1.00000           35.400000   30.000000                    1.000000   
25%      1.00000           37.800000   48.000000                    1.000000   
50%      1.00000           38.200000   64.000000                    3.000000   
75%      1.00000           38.500000   88.000000                    3.000000   
max      9.00000           40.800000  184.000000                    4.000000   

       capillary refill time  packed cell volume  
count             267.000000          270.000000  
mean                1.303371           46.300000  
std                 0.476623           10.438384  
min                 1.000000           23.000000  
25%                 1.000000           38.000000  
50%                 1.000000           45.000000  
75%                 2.000000           52.000000  
max                 3.000000           75.000000

In [10]:
# поиск моды
mod_table = pd.DataFrame()
for mod in horse_data_select.columns:
    mode = horse_data_select[mod].mode()
    mod_table[mod] = mode
mod_table

surgery?  Age  rectal temperature  pulse peripheral pulse  \
0        1    1                38.0   48.0                1   

   temperature of extremities  capillary refill time  packed cell volume  
0                         3.0                    1.0                37.0

In [11]:
horse_data_select.groupby('surgery?').count()['Age']

surgery?
1    180
2    118
Name: Age, dtype: int64

In [12]:
horse_data_select.groupby('peripheral pulse').count()['Age']

peripheral pulse
1    115
2      5
3    102
4      8
Name: Age, dtype: int64

In [13]:
horse_data_select.groupby('capillary refill time').count()['Age']

capillary refill time
1.0    188
2.0     77
3.0      2
Name: Age, dtype: int64

### Краткое описание выбранных данных:  
Всего в таблице данные по 299 лошадям, из них у 180 была операция, у 118 лечение прошло без хирургического вмешательства, по одной лошади отсутствуют данные.  
Возраст: средний 1,6 лет, самой старой лошади в базе 9 лет, самой молодой 1 год. Больше всего в базе лошадей в возрасте 1 год
Пульс: средний пульс (медиана) = 64 удара в минуту  
Средняя температура (медиана) = 38 градусов цельсия  
Температура конечностей: у 115 лошадей нормальная, у 5 лошадей - повышенная (горячая), у 102 лошадей - пониженная, у 8 - холодная  
Время наполнения капилляров: у 188 лошадей время наполнения капилляров нормальное (менее 3 сек), у 77 животных повышенное время наполнения капилляров.  
Гематокритичное число: Средний показатель (медиана) - 45, минимальный 23, максимальный 75





## Выбросы

In [14]:
# функция нахождения выбросов
def outer(df,column, norm = 0.00001):
    q1 = df[column].quantile(0.25)
    q2 = df[column].quantile(0.75)
    iqr = q2 - q1
    lower_bound = q1 - (1.5 * iqr)
    upper_bound = q2 + (1.5 * iqr)
    outers_frame = df[(df[column] <= lower_bound) | (df[column] >= upper_bound)]
    outers_frame['deviation, %'] = outers_frame.apply(lambda row: (row[column] / norm - 1) * 100, axis = 1)
    return outers_frame.sort_values(column)

In [15]:
outer(horse_data_select, 'pulse', 35)

<ipython-input-14-12e8f48827c3>:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  outers_frame['deviation, %'] = outers_frame.apply(lambda row: (row[column] / norm - 1) * 100, axis = 1)


surgery?  Age  rectal temperature  pulse peripheral pulse  \
40         2    9                39.0  150.0              NaN   
274        1    9                38.8  150.0                3   
54         1    9                38.6  160.0              NaN   
2          1    9                39.1  164.0                1   
254        1    9                38.8  184.0              NaN   

     temperature of extremities  capillary refill time  packed cell volume  \
40                          NaN                    NaN                47.0   
274                         1.0                    2.0                50.0   
54                          3.0                    1.0                38.0   
2                           4.0                    2.0                48.0   
254                         1.0                    1.0                33.0   

     deviation, %  
40     328.571429  
274    328.571429  
54     357.142857  
2      368.571429  
254    425.714286

In [16]:
outer(horse_data_select, 'rectal temperature', 37.8)

<ipython-input-14-12e8f48827c3>:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  outers_frame['deviation, %'] = outers_frame.apply(lambda row: (row[column] / norm - 1) * 100, axis = 1)


surgery?  Age  rectal temperature  pulse peripheral pulse  \
43         1    1                35.4  140.0                3   
140        2    1                36.0   42.0              NaN   
237        2    1                36.1   88.0                3   
79         1    1                36.4   98.0                3   
117        1    1                36.5   78.0              NaN   
297        1    1                36.5  100.0                3   
250        2    1                36.6   42.0                3   
98         2    1                39.6  108.0                3   
74         1    9                39.7  100.0                3   
19         1    1                39.9   72.0                1   
280        2    1                40.0   78.0                3   
53         2    1                40.3  114.0                3   
90         2    1                40.3  114.0                3   
258        1    1                40.8   72.0                3   

     temperature of extremities  capillary refill time  packed cell volume  \
43                          3.0                    2.0                57.0   
140                         NaN                    1.0                64.0   
237                         3.0                    1.0                45.0   
79                          3.0                    1.0                47.0   
117                         1.0                    1.0                34.0   
297                         3.0                    1.0                50.0   
250                         3.0                    1.0                52.0   
98                          3.0                    2.0                59.0   
74                          3.0                    2.0                48.0   
19                          1.0                    2.0                46.0   
280                         3.0                    1.0                66.0   
53                          3.0                    2.0                57.0   
90                          3.0                    2.0                57.0   
258                         3.0                    1.0                54.0   

     deviation, %  
43      -6.349206  
140     -4.761905  
237     -4.497354  
79      -3.703704  
117     -3.439153  
297     -3.439153  
250     -3.174603  
98       4.761905  
74       5.026455  
19       5.555556  
280      5.820106  
53       6.613757  
90       6.613757  
258      7.936508

In [17]:
outer(horse_data_select, 'packed cell volume', 40)

<ipython-input-14-12e8f48827c3>:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  outers_frame['deviation, %'] = outers_frame.apply(lambda row: (row[column] / norm - 1) * 100, axis = 1)


surgery?  Age  rectal temperature  pulse peripheral pulse  \
34         2    1                 NaN  104.0                3   
48         2    1                37.2   84.0                3   
3          2    1                37.3  104.0              NaN   
143        1    1                37.1   84.0                3   
150        1    1                39.3    NaN                3   

     temperature of extremities  capillary refill time  packed cell volume  \
34                          4.0                    2.0                73.0   
48                          3.0                    2.0                73.0   
3                           NaN                    2.0                74.0   
143                         3.0                    1.0                75.0   
150                         4.0                    2.0                75.0   

     deviation, %  
34           82.5  
48           82.5  
3            85.0  
143          87.5  
150          87.5

### Описание выбросов
К числовым данным, в которых могут быть потенциальные выбросы могут относится только три столбца:  
- Пульс: найдено 5 значений выбивающихся из общих данных от 150 до 184 ударов в минуту, в документации сказано "животные с болезненными поражениями или страдающие от шока кровообращения могут иметь повышенную частоту сердечных сокращений", по косвенным признакам (другим показателям), можно определить, что лошади "2" и "274" имеют проблемы со здоровьем, с чем можно связать выбросы. других лошадей необходимо исключить из выборки данных.
- Температура: 14 выбросов. Из документации: "нормальная температура составляет 37,8 ", из выборки выпали значения ниже 36,6 и выше 39,6. Отклонения не превышают 10%. По информации из сети, температура может принимать все значения, указанные в выбросах.
- Гематокритичное число: найдено 5 выбросов, отклонение от нормы в районе 85%. Из документации "Уровень повышается, когда нарушается кровообращение или когда животное становится обезвоженным." Если обратить внимание на показатель "Время наполнения капилляров", то можно увидеть, что у 4 лошадей оно повышенное, что говорит о проблемах в кровообращении, соответственно данный выброс является оправданным.  

Дополнительно: на этапе описания выявлены значения в столбце "Время наполнения капилляров" - из документации возможные значения 1 и 2, при этом в двух строках найдено значение 3, что скорей всего является ошибкой данных.


## Пропуски

In [18]:
for NAN in horse_data_select.columns:
    M = horse_data_select[NAN].isnull().sum() 
    N = M / len(horse_data_select[NAN])
 
    print(NAN, ' - ', M,  '(', round(N * 100, 1), '%)', 'пропусков')

surgery?  -  1 ( 0.3 %) пропусков
Age  -  0 ( 0.0 %) пропусков
rectal temperature  -  60 ( 20.1 %) пропусков
pulse  -  24 ( 8.0 %) пропусков
peripheral pulse  -  69 ( 23.1 %) пропусков
temperature of extremities  -  56 ( 18.7 %) пропусков
capillary refill time  -  32 ( 10.7 %) пропусков
packed cell volume  -  29 ( 9.7 %) пропусков


- surgery? : 1 пропуск -  исключить
- rectal temperature: 60 пропусков, то есть 20% от общих данных, что существенно. Заполнение: Группировка лошадей по возрасту, и температуре конечностей. Заполнение пропусков медианой
- pulse: 24 пропуска, 8% от общих данных. Заполнение: группировка по возрасту и показателю "Переферический пульс". Заполнение: медианой.
- peripheral pulse: 69 пропусков, что составляет 23,1% от общих данных, что существенно. В документации сказано показатель является субъективным. Группировка близких по значению "Пульс".
- temperature of extremities: 56 пропусков, 18,7%. Заполнение значением 2, если показатель ректальной температуры превышает температуру (37,9). Остальные данные исключить.
- capillary refill time: 32 пропуска, 10,7%. Исключить
- packed cell volume: 29 пропусков, 9,7%. Исключить

In [19]:
# surgery? 
horse_data_select = horse_data_select.dropna(subset=['surgery?'])

In [20]:
# rectal temperature
horse_data_select['rectal temperature'].fillna(horse_data_select.groupby(['Age', 'temperature of extremities'])['rectal temperature'].transform('median'), inplace = True)
horse_data_select = horse_data_select.dropna(subset=['rectal temperature'])

In [21]:
# pulse
horse_data_select['pulse'].fillna(horse_data_select.groupby(['Age', 'peripheral pulse'])['pulse'].transform('median'), inplace = True)
horse_data_select = horse_data_select.dropna(subset=['pulse'])

In [22]:
# temperature of extremities
horse_data_select[(horse_data_select['rectal temperature' ] >= 37.9)]['temperature of extremities'].fillna('2')

0      2
1      1
2      4
4      2
5      1
      ..
289    2
290    3
292    3
293    1
294    4
Name: temperature of extremities, Length: 210, dtype: object

In [23]:
horse_data_select = horse_data_select.dropna(subset=['capillary refill time'])
horse_data_select = horse_data_select.dropna(subset=['peripheral pulse'])
horse_data_select = horse_data_select.dropna(subset=['temperature of extremities'])
horse_data_select = horse_data_select.dropna(subset=['capillary refill time'])
horse_data_select = horse_data_select.dropna(subset=['packed cell volume'])

In [24]:
horse_data_select

surgery?  Age  rectal temperature  pulse peripheral pulse  \
1          2    1                38.3   40.0                1   
2          1    9                39.1  164.0                1   
5          1    1                37.9   48.0                1   
7          2    1                38.1   80.0                4   
9          1    1                38.1   66.0                3   
..       ...  ...                 ...    ...              ...   
292        1    1                38.1   78.0                3   
293        2    1                38.5   40.0                1   
295        2    1                37.2   72.0                2   
296        1    1                37.5   72.0                3   
297        1    1                36.5  100.0                3   

     temperature of extremities  capillary refill time  packed cell volume  
1                           1.0                    1.0                33.0  
2                           4.0                    2.0                48.0  
5                           1.0                    1.0                37.0  
7                           3.0                    1.0                38.0  
9                           3.0                    1.0                44.0  
..                          ...                    ...                 ...  
292                         3.0                    1.0                43.0  
293                         1.0                    1.0                37.0  
295                         3.0                    2.0                44.0  
296                         4.0                    1.0                60.0  
297                         3.0                    1.0                50.0  

[196 rows x 8 columns]

In [25]:
for NAN in horse_data_select.columns:
    M = horse_data_select[NAN].isnull().sum() 
    N = M / len(horse_data_select[NAN])
    print(NAN, ' - ', M,  '(', round(N * 100, 1), '%)', 'пропусков')

surgery?  -  0 ( 0.0 %) пропусков
Age  -  0 ( 0.0 %) пропусков
rectal temperature  -  0 ( 0.0 %) пропусков
pulse  -  0 ( 0.0 %) пропусков
peripheral pulse  -  0 ( 0.0 %) пропусков
temperature of extremities  -  0 ( 0.0 %) пропусков
capillary refill time  -  0 ( 0.0 %) пропусков
packed cell volume  -  0 ( 0.0 %) пропусков
